docs: improve locators examples and flow (#18548)
microsoft/playwright: docs/src/locators.md

> 

Reviewer: Mass comments to the snippets:
- Remove line breaks in python snippets;
- C# string options are called `NameString/NameRegex` and `HasTextString/HasTextRegex`.

await locator.click();
```

```java
Locator locator = page.frameLocator("#my-frame")
    .getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Sign in"));

locator.click();
```

```python async
locator = page.frame_locator("#my-frame")
    .get_by_role("button", name="Sign in")

await locator.click()
```

```python sync
locator = page.frame_locator("my-frame")
    .get_by_role("button", name="Sign in")

locator.click()
```

```csharp
var locator = page.FrameLocator("#my-frame")
    .GetByRole("button", new() { Name = "Sign in" });

await locator.ClickAsync();
```

### Locate by role

The [`method: Page.getByRole`] locator reflects how users and assistive technology perceive the page, for example whether some element is a button or a checkbox. When locating by role, you should usually pass the accessible name as well, so that the locator pinpoints the exact element.

For example, consider the following DOM structure.

```html
<button>Submit</button>
<input type="checkbox" id="newsletter" name="newsletter" checked>
```

You can locate each element by it's implicit role:

```js
await page.getByRole('button', { name: /submit/i }).click();

> await page.getByRole('checkbox', { checked: true, name: "newsletter" }).check();

Reviewer: Since it is already checked, we should uncheck instead :)

```suggestion
await page.getByRole('checkbox', { checked: true, name: "newsletter" }).uncheck();
```

```

```python async
await page.get_by_role("button", name=re.compile("submit", re.IGNORECASE)).click()

> await page.get_by_role("checkbox", checked=True, name="newsletter").check()

Reviewer: ```suggestion
await page.get_by_role("checkbox", checked=True, name="newsletter").uncheck()
```

```

```python sync
page.get_by_role("button", name=re.compile("submit", re.IGNORECASE)).click()

> page.get_by_role("checkbox", checked=True, name="newsletter").check()

Reviewer: ```suggestion
page.get_by_role("checkbox", checked=True, name="newsletter").uncheck()
```

```

```java
page.getByRole("button", new Page.GetByRoleOptions().setName(Pattern.compile("submit", Pattern.CASE_INSENSITIVE))).click();

> page.getByRole("checkbox", new Page.GetByRoleOptions().setChecked(true).setName("newsletter"))).check();

Reviewer: ```suggestion
page.getByRole("checkbox", new Page.GetByRoleOptions().setChecked(true).setName("newsletter"))).uncheck();
```

```

```csharp
await page.GetByRole("button", new() { Name = new Regex("submit", RegexOptions.IgnoreCase) }).ClickAsync();

> await page.GetByRole("checkbox", new() { Checked = true, Name = "newsletter" }).CheckAsync();

Reviewer: ```suggestion
await page.GetByRole("checkbox", new() { Checked = true, Name = "newsletter" }).UncheckAsync();
```

```

```csharp
await page.GetByPlaceholder("name@example.com").FillAsync("playwright@microsoft.com");
```

:::tip When to use placeholder locators
Use this locator when locating form elements that do not have labels but do have placeholder texts.
:::

### Locate by text

Find an element by the text it contains. You can match by a substring, exact string, or a regular expression when using [`method: Page.getByText`].

For example, consider the following DOM structure.

```html

> <span>Welcome John</span>

Reviewer: ```suggestion
<span>Welcome, John</span>
```

```

You can locate the element by the text it contains:

```js
await expect(page.getByText('Welcome, John')).toBeVisible();
await expect(page.getByText('Welcome, John', { exact: true })).toBeVisible();
await expect(page.getByText(/welcome, john$/i)).toBeVisible();
```

```java
assertThat(page.getByText("Welcome, John")).isVisible();
assertThat(page.getByText("Welcome, John", new Page.GetByTextOptions().setExact(true))).isVisible();

> assertThat(page.getByText(Pattern.compile("welcome john$", Pattern.CASE_INSENSITIVE))).isVisible();

Reviewer: ```suggestion
assertThat(page.getByText(Pattern.compile("welcome, john$", Pattern.CASE_INSENSITIVE))).isVisible();
```

```

```python async
await expect(page.get_by_text("Welcome, John")).to_be_visible()
await expect(page.get_by_text("Welcome, John", exact=True)).to_be_visible()

> await expect(page.get_by_text(re.compile("welcome john", re.IGNORECASE))).to_be_visible()

Reviewer: ```suggestion
await expect(page.get_by_text(re.compile("welcome, john", re.IGNORECASE))).to_be_visible()
```

```

```python sync
expect(page.get_by_text("Welcome, John")).to_be_visible()
expect(page.get_by_text("Welcome, John", exact=True)).to_be_visible()

> expect(page.get_by_text(re.compile("welcome john", re.IGNORECASE))).to_be_visible()

Reviewer: ```suggestion
expect(page.get_by_text(re.compile("welcome, john", re.IGNORECASE))).to_be_visible()
```

```

```csharp
await Expect(page.GetByText("Welcome, John")).ToBeVisibleAsync();
await Expect(page.GetByText("Welcome, John", new() { Exact: true })).ToBeVisibleAsync();

> await Expect(page.GetByText(new Regex("welcome john", RegexOptions.IgnoreCase))).ToBeVisibleAsync();

Reviewer: ```suggestion
await Expect(page.GetByText(new Regex("welcome, john", RegexOptions.IgnoreCase))).ToBeVisibleAsync();
```

```

You can also [filter by text](#filter-by-text) which can be useful when trying to find a particular item in a list.

```js
await page.getByTestId('product-item')
    .filter({ hasText: 'Playwright Book' }).click();
```
```java
page.getByTestId("product-item")
    .filter(new Locator.FilterOptions().setHasText("Playwright Book")).click();
```
```python async

> await page.get_by_test_id("product-item")

Reviewer: This is not valid python, it must either be a single line, or use a continuation character:

```py
await page.get_by_test_id("product-item") \
    .filter(has_text="Playwright Book").click()
```

I think single line is more user-friendly in this case.

    .filter(has_text="Playwright Book").click()
```
```python sync

> page.get_by_test_id("product-item")

Reviewer: Same here.

await page.GetByTestId("directions").ClickAsync();
```

By default, [`method: Page.getByTestId`] will locate elements based on the `data-testid` attribute, but you can configure it in your test config or by calling [`method: Selectors.setTestIdAttribute`].

:::tip When to use testid locators

> You can also use test ids when you choose to use the test id methodology or when you can't locate by [role](#locate-by-role) or [text](#locate-by-text).

Reviewer: ```suggestion
You should use test ids when you choose to use the test id methodology across your project.

In some rare cases it is hard to locate an element by other means, so using test id could be a one-off solution. Note that adding a test id requires changing the application in addition to the test.
```

Note that inner locator is matched starting from the outer one, not from the document root.

> ### Locate in a subtree

Reviewer: This is not really a part of "Locate elements that contain other elements". Perhaps we can do more sections?

    .click();
```

```python async
await page.get_by_test_id("product-card")
    .filter(has_text="Product 2")
    .click()

await page.get_by_test_id("product-card")
    .filter(has_text=re.compile("Product 2"))
    .click()
```

```python sync
page.get_by_test_id("product-card")
    .filter(has_text="Product 2")
    .click()

page.get_by_test_id("product-card")
    .filter(has_text=re.compile("Product 2"))
    .click()
```

```csharp
await page.GetByTestId("product-card")
    .Filter(new() { HasText = "Product 2" })
    .ClickAsync();

await page.GetByTestId("product-card")
    .Filter(new() { HasText = new Regex("Product 2") })
    .ClickAsync();
```

### Filter by another locator

Locators support an option to only select elements that have a descendant matching another locator.

For example, consider the following DOM structure:

```html
<header>
  <button>Sign in</button>
</header>

<section>
  <button>Sign in</button>
</section>
```
To select the "Sign in" button from the `section` we can first locate by the role of section and then filter by the button role.

```js
await page.getByRole('section')
    .filter({ has: page.getByRole('button', { name: 'Sign in' })

>     .click()

Reviewer: I don't think clicking on the section makes sense. Perhaps `toHaveText` is better?

Author: i thought this was clicking on the button. as it filters the button in the section or am i missing something??

```

```java
page.getByRole("section")
    .filter(new Locator.FilterOptions().setHas(page.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Sign in")))
.click()
```

```python async

> await page.get_by_role("section")

Reviewer: Same python syntax note.

    .filter(has=page.get_by_role("button", name="Sign in"))
    .click()
```

```python sync

> page.get_by_role("section")

Reviewer: Same python syntax note.

We can also assert the product card to make sure there is only one

```js
await expect(page.getByTestId('product-card')
    .filter({ has: page.getByText('Product2') }))
    .toHaveCount(1)
```

```java
await expect(page.getByTestId('product-card')
    .filter(new Locator.FilterOptions().setHas(page.getByText("Product 2")
    .hasCount(1)
```

```python async
await expect(page.get_by_role("product-card")
    .filter(has=page.get_by_role("heading", name="Product 2"))
    .to_have_count(1)
```

```python sync
expect(page.get_by_role("product-card")
    .filter(has=page.get_by_role("heading", name="Product 2"))
    .to_have_count(1)
```

```csharp
await Expect(page.GetByTestId("product-card")
    .Filter(new() { Has = page.GetByRole(AriaRole.Heading, new () { Name = "Product 2" })})
    .toHaveCountAsync(1)
```

Note that the inner locator is matched starting from the outer one, not from the document root.

## Chaining Locators

You can chain methods that create a locator, like [`method: Page.getByText`] or [`method: Locator.getByRole`], to narrow down the search to a particular part of the page. 

:::info

> You must create a locator or use filtering in order to chain locators

Author: ok, removed it as its confusing if showing a bad example so best just to show a good example

  await Expect(page.Locator("x-details")).ToContainTextAsync("Details");
  ```

## Locate within elements

You can chain methods that create a locator, like [`method: Page.getByText`] or [`method: Locator.getByRole`], to narrow down the search to a particular part of the page.

For example, consider the following DOM structure:

```html
<div data-testid='product-card'>
  <span>Product 1</span>
  <button>Buy</button>
</div>

<div data-testid='product-card'>
  <span>Product 2</span>
  <button>Buy</button>
</div>
```

Locating by role `button` will return an error due to [strictness](#strictness) as there is more than one element with this role and name. 

> In this scenario we can locate an element within another element. First we create a locator called **product** and locate the product cards by testid. We then filter by text to find the product card with the text of "Product 2". We can then use this new locator called **product** and locate the button inside that product card and click it.

Reviewer: Very nice explanation! Perhaps a numbered list would be easier to read? Up to you though.

```

However, use these methods with caution. Often times, the page might change, and locator will point to a completely different element from the one you expected. Instead, try to come up with a unique locator that will pass the [strictness criteria](#strictness).

## Strictness

Locators are strict. This means that all operations on locators that imply

> some target DOM element will throw an exception if more than one element matches a

Reviewer: `if more than one element matches a given selector`  -> `if more than one element matches`.
